cpp/include/rapidsmpf/buffer/pinned_memory_resource.hpp

```diff
@@ -0,0 +1,346 @@
+        size_t size,
+        rmm::cuda_stream_view stream,
+        std::shared_ptr<PinnedMemoryResource> mr
+    );
+
+    /**
+     * @brief Constructs a pinned host buffer and copies data into it.
+     *
+     * @param src_data Pointer to the source data to copy. If the data ptr is ordered on a
+     * stream other than @p stream, it must be synchronized with @p stream before calling
+     * this constructor. Otherwise, it will result in undefined behavior.
+     * @param size The size of the data to copy in bytes.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param mr Shared pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        void const* src_data,
+        size_t size,
+        rmm::cuda_stream_view stream,
+        std::shared_ptr<PinnedMemoryResource> mr
+    );
+
+    /**
+     * @brief Constructs a pinned host buffer by copying data from another pinned host
+     * buffer on the given stream.
+     *
+     * @param other The other pinned host buffer to copy from. If @p other is ordered
+     * on a stream other than @p stream, it must be synchronized with @p stream before
+     * calling this constructor. Otherwise, it will result in undefined behavior.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param mr Shared pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        PinnedHostBuffer const& other,
+        rmm::cuda_stream_view stream,
+        std::shared_ptr<PinnedMemoryResource> mr
+    );
+
+    /**
+     * @brief Constructs a pinned host buffer by copying data from another pinned host
+     * buffer or device buffer on the given stream.
+     *
+     * @param src The source buffer to copy from.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param mr Shared pointer to the memory resource to use for allocation and
+     * deallocation.
+     *
+     * @return A new pinned host buffer with the data copied from @p src.
+     */
+    template <typename SourceBufferT>
+        requires std::is_same_v<PinnedHostBuffer, SourceBufferT>
+                 || std::is_same_v<rmm::device_buffer, SourceBufferT>
+    static PinnedHostBuffer stream_synchronized_copy(
+        SourceBufferT const& src,
+        rmm::cuda_stream_view stream,
+        std::shared_ptr<PinnedMemoryResource> mr
+    ) {
+        if (src.size() > 0) {
+            // allocate a new buffer on the downstream stream
+            PinnedHostBuffer ret(src.size(), stream, std::move(mr));
+            // synchronize the downstream stream with upstream stream
+            cuda_stream_join(stream, src.stream());
+            RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+                ret.data(), src.data(), src.size(), cudaMemcpyDefault, stream
+            ));
+            return ret;
+        }
+        return PinnedHostBuffer(0, stream, std::move(mr));
+    }
```

Comment:
I suggest you drop the `stream` argument and just let the returned `PinnedHostBuffer` adapt the source's stream. 
Then we don't need `stream_synchronized_copy` but instead a device buffer overload:
```c++
    PinnedHostBuffer(
        rmm::device_buffer const& other,
        std::shared_ptr<PinnedMemoryResource> mr
    );
```